stream-legal-entity/legal-entity-core/src/main/java/com/backbase/stream/LegalEntitySaga.java

```diff
@@ -159,9 +143,113 @@ public Mono<LegalEntityTask> executeTask(@SpanTag(value = "streamTask") LegalEnt
+                .then(Mono.just(streamTask));
+    }
+
+    private Mono<LegalEntityTask> postServiceAgreementContacts(LegalEntityTask streamTask) {
+        LegalEntity legalEntity = streamTask.getData();
+        ServiceAgreement serviceAgreement = getServiceAgreement(legalEntity);
+        if (isEmpty(serviceAgreement.getContacts())) {
+            streamTask.info(SERVICE_AGREEMENT, PROCESS_CONTACTS, FAILED, serviceAgreement.getExternalId(), serviceAgreement.getInternalId(),
+                    "Master Service Agreement: %s does not have any Contacts defined", serviceAgreement.getExternalId());
+            return Mono.just(streamTask);
+        }
+        log.info("Creating Contacts for Service Agreement Id {}", serviceAgreement.getExternalId());
+        String externalUserId = getUserExternalId(legalEntity.getUsers());
+        if (externalUserId == null) {
+            externalUserId = getParticipantUser(serviceAgreement);
+            if (externalUserId == null) {
+                streamTask.info(LEGAL_ENTITY, PROCESS_CONTACTS, FAILED, legalEntity.getExternalId(), legalEntity.getInternalId(),
+                        "Legal Entity: %s does not have any Users", legalEntity.getExternalId());
+                return Mono.just(streamTask);
+            }
+        }
+        return contactsSaga.executeTask(createContactsTask(streamTask.getId(), legalEntity.getExternalId(), serviceAgreement.getExternalId(), externalUserId, AccessContextScope.SA, serviceAgreement.getContacts()))
+                .flatMap(contactsTask -> requireNonNull(Mono.just(streamTask)))
+                .then(Mono.just(streamTask));
+    }
+
+    private String getParticipantUser(ServiceAgreement serviceAgreement) {
+        if (!isEmpty(serviceAgreement.getParticipants())) {
+            Optional<LegalEntityParticipant> participants = serviceAgreement.getParticipants()
+                    .stream()
+                    .filter(LegalEntityParticipant::getSharingUsers)
+                    .findFirst();
+            if (participants.isPresent()) {
+                LegalEntityParticipant participant = participants.get();
+                if (!isEmpty(participant.getUsers())) {
+                    return getOptionalUserId(participant.getUsers().stream().findFirst());
+                } else if (!isEmpty(participant.getAdmins())) {
+                    return getOptionalUserId(participant.getAdmins().stream().findFirst());
+                }
+            }
+        }
+        return null;
+    }
+
+    private String getOptionalUserId(Optional<String> optionalUserId) {
+        return optionalUserId.isPresent() ? optionalUserId.get() : null;
+    }
+
+    private ContactsTask createContactsTask(String streamTaskId, String externalLegalEntityId, String externalServiceAgreementId, String externalUserId, AccessContextScope scope, List<ExternalContact> contacts) {
+        var contactData = new ContactsBulkPostRequestBody();
```

Comment:
bringing a contacts-specific DTO `ContactsBulkPostRequestBody` to the legal entity saga hurts the low coupling principle, please adjust your implementation accordingly by bringing only contact business objects to this saga and handling the contacts API specific DTOs in the contacts services layer

Reply:
This is inline to other code implementation in the following method and other places: 
    private LimitsTask createLimitsTask(LegalEntityTask streamTask, ServiceAgreement serviceAgreement, String legalEntityId, Limit limit)